sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,338 @@
+} plic_reg_info_t;
+
+/**
+ * PLIC target specific register offsets.
+ *
+ * PLIC is designed to support multiple targets, and every target has a set
+ * of its own registers. This data type is used to store PLIC target specific
+ * register offsets.
+ */
+typedef struct plic_target_reg_offset {
+  ptrdiff_t ie;        /*<< Interrupt Enable register offset. */
+  ptrdiff_t cc;        /*<< Claim/complete register offset. */
+  ptrdiff_t threshold; /*<< Threshold register offset. */
+} plic_target_reg_offset_t;
+
+/**
+ * PLIC peripheral IRQ range.
+ *
+ * PLIC IRQ source IDs are grouped together by a peripheral they belong to.
+ * Meaning that all IRQ IDs of the same peripheral are guaranteed to be
+ * consecutive. This data type is used to store IRQ ID ranges of a peripheral.
+ */
+typedef struct plic_peripheral_range {
+  dif_plic_irq_id_t first_irq_id; /*<< The first IRQ ID of a peripheral. */
+  dif_plic_irq_id_t last_irq_id;  /*<< The last IRQ ID of a peripheral. */
+} plic_peripheral_range_t;
+
+// An array of target specific set of register offsets. Every supported PLIC
+// target must have an entry in this array.
+static const plic_target_reg_offset_t
+    plic_target_reg_offsets[kDifPlicTargetCount] = {
+        [kDifPlicTargetIbex0] =
+            {
+                .ie = RV_PLIC_IE00,
+                .cc = RV_PLIC_CC0,
+                .threshold = RV_PLIC_THRESHOLD0,
+            },
+};
+
+// An array of IRQ source ID ranges per peripheral. Every peripheral supported
+// by the PLIC, must have an entry in this array.
+static const plic_peripheral_range_t
+    plic_peripheral_ranges[kDifPlicPeripheralCount] = {
+        [kDifPlicPeripheralGpio] =
+            {
+                .first_irq_id = kDifPlicIrqIdGpio0,
+                .last_irq_id = kDifPlicIrqIdGpio31,
+            },
+        [kDifPlicPeripheralUart] =
+            {
+                .first_irq_id = kDifPlicIrqIdUartTxWatermark,
+                .last_irq_id = kDifPlicIrqIdUartRxParityErr,
+            },
+        [kDifPlicPeripheralSpiDevice] =
+            {
+                .first_irq_id = kDifPlicIrqIdSpiDeviceRxF,
+                .last_irq_id = kDifPlicIrqIdSpiDeviceTxUnderflow,
+            },
+        [kDifPlicPeripheralFlashCtrl] =
+            {
+                .first_irq_id = kDifPlicIrqIdFlashCtrlProgEmpty,
+                .last_irq_id = kDifPlicIrqIdFlashCtrlOpError,
+            },
+        [kDifPlicPeripheralHmac] =
+            {
+                .first_irq_id = kDifPlicIrqIdHmacDone,
+                .last_irq_id = kDifPlicIrqIdHmacErr,
+            },
+        [kDifPlicPeripheralAlertHandler] =
+            {
+                .first_irq_id = kDifPlicIrqIdAlertHandlerClassA,
+                .last_irq_id = kDifPlicIrqIdAlertHandlerClassD,
+            },
+        [kDifPlicPeripheralNmiGen] =
+            {
+                .first_irq_id = kDifPlicIrqIdNmiGenEsc0,
+                .last_irq_id = kDifPlicIrqIdNmiGenEsc3,
+            },
+};
```

Comment:
I like the way you've structured this, it reads really nicely at the use sites and is very easy to interpret.

Reply:
Thanks, @alphan suggestion to use an array for ranges was great!

Reply:
Awesome!